### ISSUE TITLE
Nightly: thread '<unnamed>' panicked at 'couldn't compile the test', src/librustdoc/test.rs:326:13

### ISSUE THREAD

**Issue by sstangl** [rust-lang/rust#56867]
Our Rocket-based code is failing since this morning (2018-12-15). It's using the Nightly rust Docker image as a base.

The backtrace is available here: https://gitlab.com/openpowerlifting/opl-data/-/jobs/135130101

It seems to be failing in a doctest. The code wasn't changed recently; the change is likely in rustc since yesterday or the day before that.

To reproduce, you should be able to `git clone https://gitlab.com/openpowerlifting/opl-data.git` and then `cd modules/opltypes` and `cargo test`.

```
---- src/event.rs - event::Event::sd (line 67) stdout ----
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.0.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.1.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.10.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.11.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.12.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.13.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.14.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.2.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.3.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.4.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.5.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.6.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.7.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.8.rcgu.o" "/tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.9.rcgu.o" "-o" "/tmp/rustdoctestvHfhvp/rust_out" "/tmp/rustdoctestvHfhvp/rust_out.33dyzt1ekirinwy8.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs" "-L" "/builds/openpowerlifting/opl-data/target/debug/deps" "-L" "/builds/openpowerlifting/opl-data/target/debug/build/backtrace-sys-ab9883be4cf29b13/out" "-L" "/builds/openpowerlifting/opl-data/target/debug/build/ring-3ef0e65adf8c5eff/out" "-L" "/builds/openpowerlifting/opl-data/target/debug/deps" "-L" "/usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/builds/openpowerlifting/opl-data/target/debug/deps/libopltypes-4a9187ba4ef56f30.rlib" "/builds/openpowerlifting/opl-data/target/debug/deps/libstrum-d22f46e956b7b8a9.rlib" "/builds/openpowerlifting/opl-data/target/debug/deps/libserde-9db5e99214d56f64.rlib" "-Wl,--start-group" "/usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-635e0ff12adf0a0e.rlib" "/usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-6e856946e5dde60c.rlib" "/usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-5db8f2a07b5e4a3f.rlib" "/usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-67bc42cf8da35f06.rlib" "/usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-f6364059e4ec22fe.rlib" "/usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-52c0b8a76b5c5e7d.rlib" "/usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-34032f5d0b4c1746.rlib" "-Wl,--end-group" "/usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-2df59bcd1656c57e.rlib" "-Wl,-Bdynamic" "-ldl" "-lrt" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-lutil"
  = note: /tmp/rustdoctestvHfhvp/rust_out.rust_out.7rcbfp3g-cgu.9.rcgu.o: In function `std::panicking::begin_panic':
          rust_out.7rcbfp3g-cgu.9:(.text._ZN3std9panicking11begin_panic17h5b12128523b7d438E+0x40): undefined reference to `std::panicking::rust_panic_with_hook'
          collect2: error: ld returned 1 exit status
          

thread '<unnamed>' panicked at 'couldn't compile the test', src/librustdoc/test.rs:326:13
```

**Comment by stjepang** [comment]
I'm witnessing very similar linking errors in doctests of `crossbeam-channel`:

https://travis-ci.org/crossbeam-rs/crossbeam/jobs/468644578#L1083

**Comment by Centril** [comment]
cc @nagisa

**Comment by nagisa** [comment]
There are quite a few rollups in the [regression-range](https://github.com/rust-lang/rust/compare/f4a421ee3...93c2f055b) that it is hard see which one exactly could be to blame.

Here’s the excerpt of all merges. No PR jumps out as immediately obvious possible cause.

```
* 93c2f055ba4 - Auto merge of #56203 - aheart:master, r=varkor (20 hours ago) <bors>
* d652917f5f9 - Auto merge of #56847 - pietroalbini:fix-broken-toolstate, r=pietroalbini (22 hours ago) <bors>
*   44d3fb3ab20 - Auto merge of #56846 - pietroalbini:rollup, r=pietroalbini (25 hours ago) <bors>
|\  
| * 3566812b041 - (origin/pr/56846) Rollup merge of #56841 - phansch:add_various_compiletest_unittests, r=oli-obk (28 hours ago) <Pietro Albini>
| * eed96936165 - Rollup merge of #56713 - xfix:vec-test-zst-capacity, r=TimNN (28 hours ago) <Pietro Albini>
| * 8d9dc2b7493 - Rollup merge of #56710 - jethrogb:jb/sgx-target-features, r=alexcrichton (28 hours ago) <Pietro Albini>
| * b5b67298ec5 - Rollup merge of #56691 - mishaker:fix_doc_install_broken_link, r=jonas-schievink (28 hours ago) <Pietro Albini>
| * 9d50d53ec0e - Rollup merge of #56682 - alexcrichton:update-stdsimd, r=nikomatsakis (28 hours ago) <Pietro Albini>
| * adc88ac0c53 - Rollup merge of #56679 - euclio:external-doc-parse, r=estebank (28 hours ago) <Pietro Albini>
| * e92d9202699 - Rollup merge of #56677 - aelred:must-use-on-traits, r=estebank (28 hours ago) <Pietro Albini>
|/  
*   0a1b2267e45 - Auto merge of #56840 - pietroalbini:rollup, r=pietroalbini (30 hours ago) <bors>
|\  
| * ae3882c206d - (origin/pr/56840) Rollup merge of #56809 - dbrgn:permissions-ext, r=alexcrichton (32 hours ago) <Pietro Albini>
| * a352dd4dfcf - Rollup merge of #56808 - jrvidal:broken-links, r=kennytm (32 hours ago) <Pietro Albini>
| * 1f0a73039bb - Rollup merge of #56792 - phansch:add_compiletest_testsuite, r=alexcrichton (32 hours ago) <Pietro Albini>
| * 1116546e170 - Rollup merge of #56769 - dvdhrm:uefi-target, r=alexcrichton (32 hours ago) <Pietro Albini>
| * e433da7ea07 - Rollup merge of #56760 - estebank:dedup-bounds, r=oli-obk (32 hours ago) <Pietro Albini>
| * 050bb10dc3e - Rollup merge of #56758 - Manishearth:emoji-status-toolstate, r=kennytm (32 hours ago) <Pietro Albini>
| * c530e31245b - Rollup merge of #56755 - estebank:impl-trait-lt-sugg, r=cramertj (32 hours ago) <Pietro Albini>
| * 7c83737b444 - Rollup merge of #56751 - mbrubeck:hash, r=dtolnay (32 hours ago) <Pietro Albini>
| * 3a9fa9b66a7 - Rollup merge of #56747 - ljedrz:remove_box_from_target, r=zackmdavis (32 hours ago) <Pietro Albini>
| * 9544c1760c5 - Rollup merge of #56746 - pnkfelix:issue-56537-add-test-of-closure-using-region-from-containing-fn, r=nikomatsakis (32 hours ago) <Pietro Albini>
| * 846db94c667 - Rollup merge of #56738 - turboladen:fix-private_no_mangle_fns-message, r=zackmdavis (32 hours ago) <Pietro Albini>
| * 3af0cf89fae - Rollup merge of #56731 - GuillaumeGomez:ffi-doc-urls, r=Centril (32 hours ago) <Pietro Albini>
| * 63acc9ff2db - Rollup merge of #56725 - piersfinlayson:master, r=alexcrichton (32 hours ago) <Pietro Albini>
| * 9f3151abe65 - Rollup merge of #56718 - RalfJung:use-libbacktrace-printing, r=alexcrichton (32 hours ago) <Pietro Albini>
* | 747a5e514a2 - Auto merge of #56584 - davidtwco:issue-53990, r=nikomatsakis (32 hours ago) <bors>
|/  
* 7f04a646c68 - Auto merge of #56005 - GuillaumeGomez:speedup-doc-render, r=QuietMisdreavus (35 hours ago) <bors>
* 91857a3c823 - Auto merge of #56540 - oli-obk:less_const_hackery, r=varkor (2 days ago) <bors>
* bcf920fc270 - Auto merge of #56825 - alexcrichton:demangle-mem, r=nikomatsakis (2 days ago) <bors>
* 96d1334e567 - Auto merge of #56572 - kevgrasso:let_self_err_dev, r=estebank (2 days ago) <bors>
*   1897657ef09 - Auto merge of #56818 - kennytm:rollup-2, r=kennytm (2 days ago) <bors>
|\  
| * e065de2ea88 - (origin/pr/56818) Rollup merge of #56562 - pnkfelix:issue-55465-update-libc-version, r=alexcrichton (2 days ago) <kennytm>
| * 3397b79868c - Rollup merge of #56789 - alexcrichton:simd_select_bitmask, r=rkruppe (2 days ago) <kennytm>
| * 4a0ee22bc2f - Rollup merge of #56756 - tromey:Bug-56730-btree-pretty-printer, r=alexcrichton (2 days ago) <kennytm>
| * adb674ca291 - Rollup merge of #56749 - alexcrichton:adx, r=gnzlbg (2 days ago) <kennytm>
| * 123b72a05e1 - Rollup merge of #56748 - kinnison:kinnison/fix-56734, r=dtolnay (2 days ago) <kennytm>
| * 7ec1faa27a8 - Rollup merge of #56744 - ljedrz:unbox_the_children, r=matthewjasper (2 days ago) <kennytm>
| * facad1d6f24 - Rollup merge of #56709 - GuillaumeGomez:reduce-search-index, r=QuietMisdreavus (2 days ago) <kennytm>
| * 27c3631d826 - Rollup merge of #56708 - oli-obk:stability_internal_const_fn, r=alexcrichton (2 days ago) <kennytm>
| * 3e17988a537 - Rollup merge of #56702 - wesleywiser:calc_total_time_stats, r=michaelwoerister (2 days ago) <kennytm>
| * dadf7fcc2dc - Rollup merge of #56699 - nnethercote:SymbolIndex, r=oli-obk (2 days ago) <kennytm>
| * 53a2de2d549 - Rollup merge of #56695 - varkor:let-exhaustive-range, r=estebank (2 days ago) <kennytm>
| * 35fe8c92e9b - Rollup merge of #56658 - Xanewok:non-panicking-file-parser, r=petrochenkov (2 days ago) <kennytm>
| * 795f18efb8c - Rollup merge of #56637 - ollie27:rustdoc_proc_macro_local_reexport, r=QuietMisdreavus (2 days ago) <kennytm>
| * 4f0f1102bfa - Rollup merge of #56609 - michaelwoerister:unconditional-target-cpu-attr, r=alexcrichton (2 days ago) <kennytm>
* | 7d03617bab2 - Auto merge of #56568 - notriddle:master, r=alexcrichton (2 days ago) <bors>
|/  
* f4b07e0713b - (origin/pr/56816) Auto merge of #56490 - faern:add-checked-add-to-instant, r=alexcrichton (2 days ago) <bors>
* 664ede88faf - Auto merge of #56536 - alexcrichton:update-master, r=Mark-Simulacrum (2 days ago) <bors>
* 97a0bd60820 - Auto merge of #56778 - Xanewok:update-clippy, r=kennytm (3 days ago) <bors>
* 0d4f91905bf - Auto merge of #56351 - davidtwco:issue-55396-stabilize-linker-flavor, r=nagisa (3 days ago) <bors>
* 5900dae51d1 - Auto merge of #56142 - jnqnfe:osstr_lossy_example, r=alexcrichton (3 days ago) <bors>
```

**Comment by nagisa** [comment]
Observations:

* Crossbeam does not reproduce with `cargo test --doc` but reproduces with `cargo test --doc -- --test-threads=1`;
     * bisect shows a regression in #56243...
* Cannot reproduce opl-data bug with either `--test-threads=1` or without it.
     * Perhaps a request for a minimized test case would be warranted…

**Comment by pnkfelix** [comment]
triage: assigning to @nagisa

**Comment by nagisa** [comment]
@stjepang can we assume that https://github.com/crossbeam-rs/crossbeam/pull/268 fixes the issue in crossbeam?

**Comment by stjepang** [comment]
@nagisa Unfortunately, it doesn't.

That PR only affects crossbeam-skiplist, not crossbeam-channel (whose doctests are failing).

**Comment by stjepang** [comment]
@nagisa It seems the issue in crossbeam has resolved itself with a nightly update a few weeks ago.

**Comment by nikomatsakis** [comment]
Discussed in the @rust-lang/compiler meeting. Some notes:

- We'd like to separate out the crossbeam issue, which seems separate (and maybe fixed, @nagisa wants to confirm).
- @sstangl -- are you still having a problem? are you able to produce a smaller test case at all? @nagisa was unable to reproduce.

**Comment by nagisa** [comment]
Hey @sstangl, could you try seeing if the issue has also fixed itself on your project as well? If it hasn’t, then we would love a more thorough list of reproduction steps and/or description of environment in which the issue could be reproduced.

**Comment by nagisa** [comment]
Okay, I have confirmed that indeed the crossbeam issue is fixed in the most recent nightly and is not reproducible in beta, although I could not bisect what exactly fixed this because of lacking tooling support for inverted statuses.

**Comment by nagisa** [comment]
This has been fixed by 01c6ea2f37326674bf6ca64de55bf5fd90e45015.
